vhdx/src/vhdx_header.rs

```diff
@@ -0,0 +1,552 @@
+                        // Therefore, accoding to the spec, we are throwing an error.
+                        return Err(Error::UnrecognizedRegionEntry);
+                    }
+                }
+                Err(e) => {
+                    return Err(e);
+                }
+            }
+            offset += ENTRY_SIZE + 1;
+        }
+
+        if !bat_found || !mdr_found {
+            region_entries.clear();
+            return Err(Error::RegionEntryCollectionFailed);
+        }
+
+        Ok((bat_entry, mdr_entry, region_entries))
+    }
+}
+
+#[derive(Clone)]
+struct RegionEntry {
+    start: u64,
+    end: u64,
+}
+
+/// Contains the information from the header of a VHDX file
+#[derive(Clone)]
+pub struct VhdxHeader {
+    file_type_identifier: FileTypeIdentifier,
+    header_1: Header,
+    header_2: Header,
+    region_table_1: RegionTableHeader,
+    region_table_2: RegionTableHeader,
+}
+
+impl VhdxHeader {
+    /// Creates a VhdxHeader from a reference to a file
+    pub fn new(f: &mut File) -> Result<VhdxHeader> {
+        let file_type_identifier: FileTypeIdentifier = match FileTypeIdentifier::new(f) {
+            Err(e) => return Err(e),
+            Ok(r) => r,
+        };
+
+        let header_1 = Header::new(f, HEADER1_START);
+        let header_2 = Header::new(f, HEADER2_START);
+        let guid: u128 = 0; // TODO: take care of this
+        let (header_1, header_2) = VhdxHeader::update_headers(f, header_1, header_2, guid);
+
+        Ok(VhdxHeader {
+            file_type_identifier,
+            header_1,
+            header_2,
+            region_table_1: RegionTableHeader::new(f, REGION1_START)?,
+            region_table_2: RegionTableHeader::new(f, REGION2_START)?,
+        })
+    }
+
+    fn get_current_header(
+        header_1: Result<Header>,
+        header_2: Result<Header>,
+    ) -> (u8, Result<Header>, Result<Header>) {
+        let mut header1_seq_num: u64 = 0;
+        let mut header2_seq_num: u64 = 0;
+        let mut no_valid_header: bool = true;
+
+        if let Ok(ref header_1) = header_1 {
+            no_valid_header = false;
+            header1_seq_num = header_1.sequence_number;
+        }
+
+        if let Ok(ref header_2) = header_2 {
+            no_valid_header = false;
+            header2_seq_num = header_2.sequence_number;
+        }
+
+        if no_valid_header {
+            panic!("Get current header: {}", Error::NoValidHeader);
+        } else if header1_seq_num >= header2_seq_num {
+            (1, header_1, header_2)
+        } else {
+            (2, header_1, header_2)
+        }
+    }
+
+    /// _update_header() takes two header and update the noncurrent header with
+    /// the corrent one. Returns both headers as a tuple sequenced as it received
+    /// them in the parameter list.
+
+    fn _update_header(
+        f: &mut File,
+        header_1: Result<Header>,
+        header_2: Result<Header>,
+        guid: u128,
+    ) -> (Header, Header) {
```

Comment:
This should return a result, which would prevent you from using `panic!()`.
```suggestion
    ) -> Result<(Header, Header)> {
```

Reply:
Done.